lib/streamlit/delta_generator_singletons.py

```diff
@@ -0,0 +1,131 @@
+
+
+def get_bottom_dg() -> DeltaGenerator:
+    if bottom_dg is None:
+        raise RuntimeError("bottom_dg is not initialized")
+    return bottom_dg
+
+
+# The dg_stack tracks the currently active DeltaGenerator, and is pushed to when
+# a DeltaGenerator is entered via a `with` block. This is implemented as a ContextVar
+# so that different threads or async tasks can have their own stacks.
+def get_default_dg_stack() -> tuple[DeltaGenerator, ...]:
+    if main_dg is None:
+        raise RuntimeError("main_dg is not set")
+
+    return (main_dg,)
+
+
+_T = TypeVar("_T")
+
+
+class ContextVarWithLazyDefault(Generic[_T]):
+    def __init__(self, name: str, *, default: Callable[[], _T]):
+        self._name = name
+        self._default = default
+        self._context_var: ContextVar[_T] | None = None
+
+    def _init_context_var(self) -> None:
+        self._context_var = ContextVar(self._name, default=self._default())
+
+    def get(self) -> _T:
+        if self._context_var is None:
+            self._init_context_var()
+        return self._context_var.get()  # type: ignore[union-attr]
+
+    def set(self, value: _T) -> Token[_T]:
+        if self._context_var is None:
+            self._init_context_var()
+        return self._context_var.set(value)  # type: ignore[union-attr]
+
+    def reset(self, token: Token[_T]) -> None:
+        if self._context_var is None:
+            self._init_context_var()
+        self._context_var.reset(token)  # type: ignore[union-attr]
+
+    def __hash__(self) -> int:
+        if self._context_var is None:
+            self._init_context_var()
+        return self._context_var.__hash__()
+
+
+# we don't use the default factory here because `main_dg` is not initialized when this
+# module is imported. This is why we have our own ContextVar wrapper.
+context_dg_stack: ContextVarWithLazyDefault[tuple[DeltaGenerator, ...]] = (
+    ContextVarWithLazyDefault("context_dg_stack", default=lambda: (get_main_dg(),))
+)
+
+
+def get_last_dg_added_to_context_stack() -> DeltaGenerator | None:
+    """Get the last added DeltaGenerator of the stack in the current context.
+
+    Returns None if the stack has only one element or is empty for whatever reason.
+    """
+    current_stack = context_dg_stack.get()
+    # If set to "> 0" and thus return the only delta generator in the stack -
+    # which logically makes more sense -, some unit tests fail.
+    # It looks like the reason is that they create their own main delta generator
+    # but do not populate the dg_stack correctly. However, to be on the safe-side,
+    # we keep the logic but leave the comment as shared knowledge for whoever will look
+    # into this in the future.
+    if len(current_stack) > 1:
+        return current_stack[-1]
+    return None
+
+
+# stubs for functions that are required by some Mixins. Defined here to avoid circular
+# imports.
+create_status_container: Callable[..., StatusContainer]
+create_dialog: Callable[..., Dialog]
```

Comment:
at first sight, I'm not too big of a fan of how this aspect is solved... but need to think about this a bit more

Reply:
Yeah I know what you mean. But other solutions seemed to be way more involved like
1. having proper interfaces to resolve the circular import in a completely different way
2. I have played around with dynamic inheritance like `Dialog(type(parent))` to not import `DeltaGenerator` for the inheritance and while this worked, it was not great either and also we would have lost auto-completion on `st.dialog.*` and `st.status.*` because it would not have been a `DeltaGenerator` anymore.
3. Perhaps refactoring `DeltaGenerator` so that it is used in a composite-way instead of inheritance. This would actually be the cleanest solution going forward, but likely would involve some more refactoring as it might involve having interfaces for the autocompletion aspect etc. as in point 2

Since the `delta_generator_singletons.py` is used for breaking the circular imports, I don't think it is a terrible solution and we could go ahead with it; since I think it is still better than having the circular imports. Step by step 🙂

Reply:
Ok, I thought about for a while. I think what makes it confusing for me is how the module is used as basis for the singleton. I think I would feel better with a class-based singelton based on the same concept as `Runtime`. 

`delta_generator_singletons.py`:

```python
....

class DeltaGeneratorSingleton:
    _instance: DeltaGeneratorSingleton | None = None

    @classmethod
    def instance(cls) -> DeltaGeneratorSingleton:
        """Return the singleton DeltaGeneratorSingleton instance. Raise an Error if the
        Runtime hasn't been created yet.
        """
        if cls._instance is None:
            raise RuntimeError("DeltaGeneratorSingleton hasn't been created!")
        return cls._instance

    def __init__(
        self,
        delta_generator_cls: type[DeltaGenerator],
        status_container_cls: type[StatusContainer],
        dialog_container_cls: type[Dialog],
    ):
        """Registers and initializes all delta-generator classes.

        Parameters
        ----------

        delta_generator_cls : type[DeltaGenerator]
            The main DeltaGenerator class.
        status_container_cls : type[StatusContainer]
            The delta-generator class that is used as return value for `st.status`.
        dialog_container_cls : type[Dialog]
            The delta-generator class used is used as return value for `st.dialog`.

        Raises
        ------
        RuntimeError
            If the DeltaGeneratorSingleton instance already exists.
        """
        if DeltaGeneratorSingleton._instance is not None:
            raise RuntimeError("DeltaGeneratorSingleton instance already exists!")
        DeltaGeneratorSingleton._instance = self

        self._main_dg = delta_generator_cls(root_container=_RootContainer.MAIN)
        self._sidebar_dg = delta_generator_cls(
            root_container=_RootContainer.SIDEBAR, parent=self._main_dg
        )
        self._event_dg = delta_generator_cls(
            root_container=_RootContainer.EVENT, parent=self._main_dg
        )
        self._bottom_dg = delta_generator_cls(
            root_container=_RootContainer.BOTTOM, parent=self._main_dg
        )
        self._status_container_cls = status_container_cls
        self._dialog_container_cls = dialog_container_cls

    @property
    def main_dg(self) -> DeltaGenerator:
        return self._main_dg

    @property
    def sidebar_dg(self) -> DeltaGenerator:
        return self._sidebar_dg

    @property
    def event_dg(self) -> DeltaGenerator:
        return self._event_dg

    @property
    def bottom_dg(self) -> DeltaGenerator:
        return self._bottom_dg

    @property
    def status_container_cls(self) -> type[StatusContainer]:
        return self._status_container_cls

    @property
    def dialog_container_cls(self) -> type[Dialog]:
        return self._dialog_container_cls


def get_delta_generator_singleton() -> DeltaGeneratorSingleton:
    """Return the singleton Runtime instance. Raise an Error if the
    Runtime hasn't been created yet.
    """
    return DeltaGeneratorSingleton.instance()

```

And initialize the singleton in `__init__.py`:

```python
_dg_singletons = DeltaGeneratorSingleton(
    delta_generator_cls=DeltaGenerator,
    status_container_cls=StatusContainer,
    dialog_container_cls=Dialog,
)

_main = _dg_singletons.main_dg
sidebar = _dg_singletons.sidebar_dg
_event = _dg_singletons.event_dg
_bottom = _dg_singletons.bottom_dg
```

This feels more aligned with other singelton's in Streamlit, allows better docstring based documentation, and just feels slightly easier to understand compared to a collection of module-level variables.

Reply:
Also, I would love if we find a better name independent of delta generator for this construct: e.g. `AppManager`, `AppLayout`, `AppCreator`, `RootContainer`, ...

Reply:
yeah this is how I had it originally in the [spec](https://www.notion.so/snowflake-corp/TechSpec-Refactor-DeltaGenerator-727d231a3f5244cfbcde5dfe6ed4e902?pvs=4#0a9c8134953a4a5fb51de4ad98c0726c), but then while implementing it decided against it because you don't need that kind of construct in Python to achieve the same goal and the difference to Runtime etc. is that those are part of our internal agreed-on APIs. So I found the whole "you have to create a class and check for instance" etc. for the same thing overkill. But if you feel more strongly about it, I will update it again 👍

I will also think about the name, although we might want to do that in a followup PR and brainstorm the name with the team.

Reply:
Oh, I see.. that was my initial expectation when thinking about the singelton. Maybe because I saw it in your spec. But I think I would prefer a class based singelton since at least for me it makes it easier to understand and reason about (+ benefits in documenting it).

Reply:
I have updated it to be a class-based singleton 👍